### DISCUSSION TITLE
Save solver state

### DISCUSSION THREAD

**Discussion by rybickibartek** [google/or-tools#2427] [CP-SAT questions] [answered]
**What language and solver does this apply to?**
Python, CP-SAT

**Describe the problem you are trying to solve.**
I would like to store solver state between sessions / executions.

I am running CP-SAT on an optimization model. I stop it at some point and add some new constraints
- set of variables is the same 
- some variables are fixed with values [these are only constraints added]
- best found solution is feasible
- all variables have hint [last found solution]

I would like a solver to start from 'posiiton' in which I stopped it - to do not check solutions already seen in previous session(s) / execution(s).

Is it possible to be done? Or maybe it is already implemented?

**Comment by lperron** [comment, marked as answer]
The solve() itself is stateless.
You can store the last solution, and restart from it using hints in the model.

See https://github.com/google/or-tools/blob/stable/ortools/sat/docs/model.md#solution-hinting

**Reply by rybickibartek** [reply]
This is exactly what I am doing. But does it prevents solver from checking already visited partial solutions which leads either infeasible solutions or solutions with are worst then best already observed solution?

**Reply by pramitbiswas** [reply]
@lperron Can you please update the link, or provide relevant information?

**Reply by lperron** [reply]
Done

**Comment by lperron** [comment]
No, it does not.
Still, it is the best you can do actually.
Laurent Perron | Operations Research | lperron@google.com | (33) 1 42 68 53
00



Le lun. 8 mars 2021 à 16:17, rybickibartek <notifications@github.com> a
écrit :

**Reply by rybickibartek** [reply]
Ok great, thanks for help!
